### ISSUE TITLE
Fix stats

### ISSUE THREAD

**Issue by timmaxw** [rethinkdb/rethinkdb#3883]
The `rethinkdb.stats` system table is currently broken in `raft`. We need to fix it. The main parts of this are:

* Change how the stats system finds table metadata

* Make sure that the perfmon names that the new Raft system publishes are the same as the perfmon names that the stats system looks for

**Comment by VeXocide** [comment]
@coffeemug @danielmewes -- we currently bail out if we fail to contact at least one server for every table, is that acceptable?

**Comment by coffeemug** [comment]
I think this isn't very useful. I may be running a cluster with two dozen tables, piping my stats feed into a monitoring tool. If a part of the cluster goes offline, I still want to see the stats for the rest of the tables -- I don't want the whole monitoring system to go dark. I think we should fix this as part of the refactor.

I think we should definitely skip servers that aren't responding and move on. We shouldn't abort, even if none of the servers for a particular table are available.

There is a question of whether we should add an entry for inaccessible servers/tables to the `stats` table (to inform the user that we couldn't read the stats for those particular resources). I think we should, but would be interesting to know what @danielmewes @AtnNn @larkost think.

**Comment by timmaxw** [comment]
There's a closely related issue: What do we show in the `table_config` system table if we can't retrieve the config for the table because all of the servers are inaccessible?

**Comment by danielmewes** [comment]
In 2.0 we currently do the following:
* for stats entries that are for a specific server, we generate a document like this:

    ```
{
    "error": "Timed out. Unable to retrieve stats." ,
    "id": [
        "server" ,
        "ac11075b-0523-4a81-9f1f-583425e1bef9"
    ] ,
    "server": "starearth_or0"
}
```
    This includes server and table_server documents.
* For documents that are not server-specific (i.e. `table`, `cluster`), we ignore the numbers from the unavailable servers and also don't show any error.

I think we should keep that behavior if possible.

I suggest that in `table_config`, we generate a similar document as in case 1 with an `error` field.

**Comment by VeXocide** [comment]
@danielmewes -- the `"Timed out."` still works as before as it times out fetching the statistics. We first have to figure out which servers to contact per table to begin with, which is done using the configs and those have to be fetched as well.

**Comment by danielmewes** [comment]
For the stats table, can't we just contact all connected servers once and have them send us stats for all their tables which we can then merge? Or would that be a big change from the current implementation?

Does the non-authorative table metadata have a snapshot of servers for a table that we could use to generate table_server and server entries for missing servers?

**Comment by timmaxw** [comment]
The non-authoritative table metadata records the table's UUID, name, database UUID, and primary key.

**Comment by timmaxw** [comment]
I think that if we can't fetch a table's config, we should report an error document for that table, but not report any table/server pair documents for that table.

**Comment by danielmewes** [comment]
> I think that if we can't fetch a table's config, we should report an error document for that table, but not report any table/server pair documents for that table.

That sounds good enough. :+1:

**Comment by coffeemug** [comment]
Sounds good to me too.

**Comment by larkost** [comment]
My 2¢ is that we definitely should show missing tables as missing (I think there is consensus on that), and if reasonably possible show a guess about what servers are involved. From the admin's perspective knowing that a table is missing is very important, but the next thing I want to know is what server I have to kick to get it back up.

If we can't reasonably give that information, then we can't. But that will be a point of frustration.

**Comment by VeXocide** [comment]
For the record, this is in CR 2851 by @timmaxw.

**Comment by VeXocide** [comment]
The issue of handling the failure to contact any servers on which a table lives means changing the interface of `table_meta_client::list_configs` and fixing all uses. I'll open a separate issue for that.

**Comment by danielmewes** [comment]
:+1:

**Comment by VeXocide** [comment]
Merged into `raft` via commit 8d2f46b5da1b560c275b570df2ecbea3891893df.
